### ISSUE TITLE
Database locking issues

### ISSUE THREAD

**Issue by keywordnew** [borgbase/vorta#1081]
Hi folks, I used the template but may have omitted useful detail. If so, let me know what to look for.

**Describe the bug**
Vorta crashes near the end of scheduled backups, apparently while in the prune stage.

**To Reproduce**
Seems to occur at the pruning step of scheduled backups. See log extract below. 
Crash does not happen with every prune.

**Environment:**
- OS: MacOS 10.15.7
- Vorta version: 0.7.8
- Installed from: https://github.com/Homebrew/homebrew-cask/blob/HEAD/Casks/vorta.rb
- borg version: 1.1.17

**Additional context**

```
2021-10-18 10:24:56,037 - vorta.borg.borg_thread - INFO - Keeping archive: REDACTED, NOT ISOLATED TO ONE REPO
2021-10-18 10:24:56,151 - vorta.keyring.abc - DEBUG - No module named 'secretstorage'
2021-10-18 10:24:56,152 - vorta.borg.borg_thread - DEBUG - Using VortaDarwinKeyring keyring to store passwords.
2021-10-18 10:24:56,160 - vorta.borg.borg_thread - INFO - Running command /usr/local/bin/borg list --info --log-json --json /Users/REDACTED
2021-10-18 10:24:56,641 - root - CRITICAL - Uncaught exception, file a report at https://github.com/borgbase/vorta/issues/new
Traceback (most recent call last):
  File "peewee.py", line 3144, in execute_sql
sqlite3.OperationalError: database is locked

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "borg/borg_thread.py", line 273, in run
  File "borg/list_repo.py", line 43, in process_result
  File "peewee.py", line 6603, in delete_instance
  File "peewee.py", line 1907, in inner
  File "peewee.py", line 1978, in execute
  File "peewee.py", line 2474, in _execute
  File "peewee.py", line 3157, in execute
  File "peewee.py", line 3151, in execute_sql
  File "peewee.py", line 2917, in __exit__
  File "peewee.py", line 190, in reraise
  File "peewee.py", line 3144, in execute_sql
peewee.OperationalError: database is locked
```

IMO this issue is similar to #1032. Opening separate issue because that one doesn't specify a link to pruning.

Regular crashes for me seems to have started when I upgraded from 0.7.4 to 0.7.6. Before this point, Vorta ran predictably. After this point, Vorta now crashes at least once daily. I can't say that the cause has been consistent, I only have the most recent logs. 

In logs for the past 24 hours, the string "root - CRITICAL - Uncaught exception" has shown up 4 times. This seems to correspond with the frequency of crashes.

Another more minor issue that seems to have started at the same time, is that selecting a Profile from the dropdown makes the app freeze for several seconds. In the past, it occasionally permanently rainbow wheeled, but now the freezing reliably recovers.

**Comment by m3nu** [comment]
Using this different sqlite engine would solve this. Since we don't have a DB lock in the app.

http://docs.peewee-orm.com/en/latest/peewee/playhouse.html#sqliteq

**Comment by bastiencyr** [comment]
In a very recent peewee version you have another way to ensure thread safety  : http://docs.peewee-orm.com/en/latest/peewee/database.html?highlight=thread#thread-safety-and-multiple-databases 
But we need pewee==3.14.6 at least.

**Comment by m3nu** [comment]
Just tried the `SqliteQueueDatabase` approach, but had many issues because the order of DB queries isn't guaranteed any more. So can't use that.

Will now try using a global lock just for `BorgJob`. Since that's where both errors were reported.

**Comment by bastiencyr** [comment]
Is it temporary ? 
If you use a global lock, so the jobsmanager will not be able to run concurrent job. 
Where do you plan to put the global lock exactly ?

**Comment by m3nu** [comment]
I ended up putting a second global lock (you already added one for keyring operations) for DB operations ONLY. Not for any other ops. This shouldn't influence JobsManager much.

My PR: https://github.com/borgbase/vorta/pull/1109/files

Do you think this will address the issue while giving us the benefits of JobsManager? I didn't need locking very often before. So I'm not too familiar here.

Also made a small change to use a context manager instead of acquire/release. Seems more readable.

**Comment by m3nu** [comment]
My understanding is that only ONE thread will run the locked code block at the same time. Be it keyring or DB. We only lock the critical blocks.

**Comment by keywordnew** [comment]
> m3nu closed this in #1109 2 days ago

🎉

Looking forward to re-enabling automatic pruning. Thank you :-)

**Comment by m3nu** [comment]
Thanks for reporting, @keywordnew! This error is too rare to reproduce, but I hope this new lock will avoid multiple threads from accessing the DB at the same time.
